src/dali/plugin/input/rest/binance_com.py

```diff
@@ -0,0 +1,1295 @@
+_TOTAL: str = "total"
+_TOTALFEE: str = "totalFee"
+_TOTALNUM: str = "totalNum"
+_TYPE: str = "type"
+_TXID: str = "txid"  # CCXT doesn't capitalize I
+_TXNTYPE: str = "txnType"
+_UPDATETIME: str = "updateTime"
+_USERNAME: str = "userName"
+_WITHDRAWAL: str = "withdrawal"  # CCXT only variable
+
+# Time period constants
+_NINETY_DAYS_IN_MS: int = 7776000000
+_THIRTY_DAYS_IN_MS: int = 2592000000
+_ONE_DAY_IN_MS: int = 86400000
+_MS_IN_SECOND: int = 1000
+
+# Record limits
+_DEPOSIT_RECORD_LIMIT: int = 1000
+_DIVIDEND_RECORD_LIMIT: int = 500
+_DUST_TRADE_RECORD_LIMIT: int = 100
+_INTEREST_SIZE_LIMIT: int = 100
+_MINING_PAGE_LIMIT: int = 200
+_TRADE_RECORD_LIMIT: int = 1000
+_WITHDRAWAL_RECORD_LIMIT: int = 1000
+
+# Types of Binance Dividends
+_BNBVAULT = "BNB Vault"
+_ETHSTAKING = "ETH 2.0 Staking"
+_FLEXIBLESAVINGS = "Flexible Savings"
+_LAUNCHPOOL = "Launchpool"
+_LOCKEDSAVINGS = "Locked Savings"
+_LOCKEDSTAKING = "Locked Staking"
+_SOLOAIRDROP = "SOLO airdrop"
+
+_AIRDROP_LIST = [_SOLOAIRDROP]
+_INTEREST_LIST = [_FLEXIBLESAVINGS, _LOCKEDSAVINGS]
+_STAKING_LIST = [_ETHSTAKING, _LOCKEDSTAKING, _BNBVAULT, _LAUNCHPOOL]
+
+
+class _ProcessAccountResult(NamedTuple):
+    in_transactions: List[InTransaction]
+    out_transactions: List[OutTransaction]
+    intra_transactions: List[IntraTransaction]
+
+
+class _Trade(NamedTuple):
+    base_asset: str
+    quote_asset: str
+    base_info: str
+    quote_info: str
+
+
+class InputPlugin(AbstractInputPlugin):
+
+    __BINANCE_COM: str = "Binance.com"
+
+    def __init__(
+        self,
+        account_holder: str,
+        api_key: str,
+        api_secret: str,
+        native_fiat: str,
+        username: Optional[str] = None,
+    ) -> None:
+
+        super().__init__(account_holder, native_fiat)
+        self.__logger: logging.Logger = create_logger(f"{self.__BINANCE_COM}/{self.account_holder}")
+        self.__cache_key: str = f"{self.__BINANCE_COM.lower()}-{account_holder}"
+        self.client: binance = binance(
+            {
+                "apiKey": api_key,
+                "secret": api_secret,
+            }
+        )
+        self.username = username
+
+        # We have to know what markets and algos are on Binance so that we can pull orders using the market
+        self.markets: List[str] = []
+        self.algos: List[str] = []
+
+        # We will have a default start time of July 13th, 2017 since Binance Exchange officially launched on July 14th Beijing Time.
+        self.start_time: datetime = datetime(2017, 7, 13, 0, 0, 0, 0)
+        self.start_time_ms: int = int(self.start_time.timestamp()) * _MS_IN_SECOND
+
+    def cache_key(self) -> Optional[str]:
+        return self.__cache_key
+
+    @staticmethod
+    def _rp2timestamp_from_ms_epoch(epoch_timestamp: str) -> str:
+        rp2_time = datetime.fromtimestamp((int(epoch_timestamp) / _MS_IN_SECOND), timezone.utc)
+
+        # RP2 Timestamp has a space between the UTC offset and seconds

```

Comment:
Ouch, this is a bug: thanks for pointing it out. My intention was to support ISO8601, but somehow I got convinced there was a space between time and timezone. Upon rereading the standard, it's clear there is no space. I'll fix it.

Reply:
The extra space in timestamp bug is now fixed: https://github.com/eprbell/dali-rp2/commit/8b92612c8c60b2b0c5ef4c49259372614d4f00e1